blog/2020/05/invalidations.md

```diff
@@ -0,0 +1,733 @@
+The good news is that these counts reveal that more than half of all invalidations will likely be fixed by "automated" means.
+However, it appears that there will need to be a second round in which package developers inspect individual invalidations to determine what, if anything, can be done to remediate them.
+
+## Fixing invalidations
+
+You may have noticed that two packages, `Example` and `Revise`, trigger far fewer invalidations that the rest of the packages in our analysis.
+`Example` is quite trivial, but `Revise` and its dependencies are quite large.
+How does it avoid this problem?
+First, Revise does not extend very many Base methods;
+most of its methods are for functions it "owns," and the same is true for its dependencies.
+Second, in the closing days of Julia 1.5's merge window,
+Revise (and Julia) underwent a process of tracking down invalidations and eliminating them;
+for comparison, on Julia 1.4, Revise triggers more than a 1000 non-unique invalidations.
+The success of this effort gives one hope that other packages too may one day have fewer invalidations.
+
+As stated above, there is reason to hope that most of the invalidations marked as "ambiguous" will be fixed by changes to Julia's compiler.
+Here our focus is on those marked "more specific," since those are cases where it is harder to imagine a generic fix.
+
+### Fixing type instabilities
+
+In engineering Julia and Revise to reduce invalidations, at least two cases were fixed by resolving type-instabilities.
+For example, one set of invalidations happened because CodeTracking, a dependency of Revise's, defines new methods for `Base.PkgId`.
+It turns out that this triggered an invalidation of `_tryrequire_from_serialized`, which is used to load packages.
+Fortunately, it turned out to be an easy fix: one section of `_tryrequire_from_serialized` had a passage
+
+```
+for M in mod::Vector{Any}
+    if PkgId(M) == modkey && module_build_id(M) === build_id
+        return M
+    end
+end
+```
+
+and since `M` had type `Any`, the compiler couldn't predict which version of `PkgId` would be called.
+It sufficed to add
+
+```
+    M = M::Module
+```
+
+immediately after the `for` statement to fix the problem.
+Not only does this fix the invalidation, but it lets the compiler generate better code.
+
+The other case was a call from `Pkg` of `keys` on an AbstractDict of unknown type
+(due to partial specialization).
+Replacing `keys(dct)` with `Base.KeySet(dct)` (which is the default return value of `keys`) eliminated a very consequential invalidation, one that triggered seconds-long latencies in the next `Pkg` command after loading Revise.
```

Comment:
We do get zero-cost abstraction when inference works. However, this seems to be a case where it doesn't. As of today I have better tools, so maybe I'll actually revert that change and fix it "properly." But hold on to your hat, you'll see this takes a lot more sophistication than the cheap version.

For reference, we're investigating https://github.com/JuliaLang/Pkg.jl/blob/fff698d3d7d90a55431ad2f4fa0af9719d899954/ext/TOML/src/print.jl#L53

```julia
julia> using MethodAnalysis

julia> using Pkg

julia> f = Pkg.TOML._print
_print (generic function with 2 methods)

julia> direct_backedges(f)
2-element Array{Any,1}:
                    MethodInstance for _print(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Dict{String,Array{Dict{String,Dates.DateTime},1}}) => MethodInstance for #print#10(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(Pkg.TOML.print), ::Base.GenericIOBuffer{Array{UInt8,1}}, ::Dict{String,Array{Dict{String,Dates.DateTime},1}})
 MethodInstance for _print(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Dict{String,Array{Dict{String,Dates.DateTime},1}}, ::Array{String,1}) => MethodInstance for _print(::Base.GenericIOBuffer{Array{UInt8,1}}, ::Dict{String,Array{Dict{String,Dates.DateTime},1}})
```

Hmm, that's curious, all the calls are concretely typed for the dict, which makes it a mystery why the `keys` call wasn't inferred. Since this is a keyword method, maybe we should check the kw function. Since it turns out we get a hit, let's be selective and filter for them:

```julia
julia> fkw = Core.kwfunc(f)
_print##kw (generic function with 2 methods)

julia> bes = filter(pr->!isconcretetype(pr.first.specTypes), direct_backedges(fkw))
2-element Array{Any,1}:
                    MethodInstance for (::Pkg.TOML.var"#_print##kw")(::NamedTuple{(:sorted,),Tuple{Bool}}, ::typeof(Pkg.TOML._print), ::IOStream, ::AbstractDict) => MethodInstance for #print#10(::Base.Iterators.Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:sorted,),Tuple{Bool}}}, ::typeof(Pkg.TOML.print), ::IOStream, ::AbstractDict)
 MethodInstance for (::Pkg.TOML.var"#_print##kw")(::NamedTuple{(:sorted,),Tuple{Bool}}, ::typeof(Pkg.TOML._print), ::IOStream, ::AbstractDict, ::Array{String,1}) => MethodInstance for (::Pkg.TOML.var"#_print##kw")(::NamedTuple{(:sorted,),Tuple{Bool}}, ::typeof(Pkg.TOML._print), ::IOStream, ::AbstractDict)

julia> using AbstractTrees

julia> print_tree(bes[1].first, 10)
MethodInstance for (::var"#_print##kw")(::NamedTuple{(:sorted,),Tuple{Bool}}, ::typeof(_print), ::IOStream, ::AbstractDict)
└─ MethodInstance for #print#10(::Pairs{Symbol,Bool,Tuple{Symbol},NamedTuple{(:sorted,),Tuple{Bool}}}, ::typeof(print), ::IOStream, ::AbstractDict)
   └─ MethodInstance for (::var"#print##kw")(::NamedTuple{(:sorted,),Tuple{Bool}}, ::typeof(print), ::IOStream, ::AbstractDict)
      ├─ MethodInstance for (::var"#28#32")(::IOStream)
      │  ├─ MethodInstance for #open#284(::Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(open), ::var"#28#32", ::String, ::Vararg{Any,N} where N)
      │  ├─ MethodInstance for #open#284(::Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(open), ::var"#28#32", ::String, ::Vararg{String,N} where N)
      │  │  └─ MethodInstance for open(::var"#28#32", ::String, ::String)
      │  │     ├─ MethodInstance for #bind_artifact!#24(::Linux, ::Nothing, ::Bool, ::Bool, ::typeof(bind_artifact!), ::String, ::String, ::SHA1)
 ⋮
```
So we've identified `bind_artifact!` as one likely culprit (because the dict appears after that).

Using Cthuhlu one can discover that the fallback definition https://github.com/JuliaLang/Pkg.jl/blob/fff698d3d7d90a55431ad2f4fa0af9719d899954/src/Artifacts.jl#L632 is one culprit, it should be `Dict{String,Any}()`.

However, that's not the only culprit: the other is https://github.com/JuliaLang/julia/issues/15276.

And, it involves not just `bind_artifact!` but also `load_telemetry_file` and `get_auth_header`. 

Especially since it involves 15276, I'm just not sure this is blog worthy. Thoughts?

Reply:
https://github.com/JuliaLang/Pkg.jl/pull/1824

Reply:
Thanks for the extra in-depth explanation!

Would using `TOML.print(io, convert(Dict{String,Any}, dict), sorted=true)` everywhere in Pkg work too?  Since Pkg knows the concrete type with which it is working, I think it makes sense to type-stabilize at the API boundary.  (This way, `keys` in TOML is still generic.)

By the way, my point on "zero-cost abstraction" is not only about `KeySet` (but `KeySet` is a very good example). It's more about balancing abstraction and load-time optimization in general. Are we going to have to restrict abstraction for reducing load-time? Or would it turn out to be not much of a concern and even help us find/implement a better abstraction? `reduce_empty` is a bit hopeful example in this direction.

Reply:
> Would using TOML.print(io, convert(Dict{String,Any}, dict), sorted=true) everywhere in Pkg work too? 

Yes, though that wasn't clear to me at the outset. But the PR I submitted is a better solution since it fixes the root cause.

Reply:
I feel what is better is not entirely objective here (e.g., depending less on compiler implementation details may be preferable in some situations).

Anyway, as I said, `KeySet` is just one concrete example that load-time optimization can potentially be in conflict with the abstraction of the interface.  I don't think addressing one example solves the root problem.  Putting this as an open hard problem in the blog post seems to be a good balance to me. (Though I'm aware maybe I have an extreme point of view for this. Maybe the post as-is is fine.)